obspy/io/rg16/core.py

```diff
@@ -0,0 +1,238 @@
+        traces or that the instruments were deployed with the gold contact
+        terminals facing north. If this parameter is used, it will map the
+        components to Z, N, and E (if 3C) as well as correct the polarity for
+        the vertical component.
+    :type contacts_north: bool
+    :return: An ObsPy :class:`~obspy.core.stream.Stream` object.
+
+    """
+    if not is_rg16(filename):
+        raise ValueError('read_fcnt was not passed a Fairfield RG 1.6 file')
+    # get timestamps
+    time1 = UTCDateTime(starttime).timestamp if starttime else 0
+    time2 = UTCDateTime(endtime).timestamp if endtime else BIG_TS
+    # read general header information
+    gheader = _read_block(filename, general_header_block)
+    # byte number channel sets start at in file
+    chan_set_start = (gheader['num_additional_headers'] + 1) * 32
+    # get the byte number the extended headers start
+    eheader_start = (gheader['channel_sets']) * 32 + chan_set_start
+    # read trace headers
+    ex_headers = gheader['extended_headers'] + gheader['external_headers']
+    # get byte number trace headers start
+    theader_start = eheader_start + (ex_headers * 32)
+    # get traces and return stream
+    traces = _make_traces(filename, theader_start, gheader, head_only=headonly,
+                          starttime=time1, endtime=time2, merge=merge,
+                          standard_orientation=contacts_north)
+    return Stream(traces=traces)
+
+
+@open_file
+def is_rg16(filename, **kwargs):
+    """
+    Determine if a file or buffer contains an rg16 file.
+
+    :param filename: A path to the file or a buffer of an opened file.
+    :type filename: str, buffer
+    :return: bool
+    """
+    try:
+        filename.seek(0)
+        sample_format = _read(filename, 2, 2, 'bcd')
+        manufacturer_code = _read(filename, 16, 1, 'bcd')
+        version = _read(filename, 42, 2, None)
+    except ValueError:  # if file too small
+        return False
+    con1 = version == b'\x01\x06' and sample_format == 8058
+    return con1 and manufacturer_code == 20
+
+
+# ------------ helper functions for formatting specific blocks
+
+
+def _make_traces(fi, data_block_start, gheader, head_only=False,
+                 starttime=None, endtime=None, merge=False,
+                 standard_orientation=False):
+    """ make obspy traces from trace blocks and headers """
+    traces = []  # list to store traces
+    trace_position = data_block_start
+    while True:  # read traces until parser falls of the end of file
+        try:
+            theader = _read_block(fi, trace_header_block, trace_position)
+        except ValueError:  # this is the end, my only friend, the end
+            break
+        # get stats
+        stats = _make_stats(theader, gheader, standard_orientation)
+        # expected jump to next start position
+        jumps = stats.npts * 4 + theader['num_ext_blocks'] * 32 + 20
```

Comment:
Any chance this could be zero with corrupt data? It would then be trapped in an infinite loop.

Reply:
Good point, I will add an assert that jumps != 0. It shouldn't ever happen but if it does it would cause a big problem.